parquet/src/arrow/async_reader/metadata.rs

```
@@ -71,7 +71,10 @@ impl<F: MetadataFetch> MetadataLoader<F> {
         let suffix_len = suffix.len();
 
         let mut footer = [0; 8];
-        footer.copy_from_slice(&suffix[suffix_len - 8..suffix_len]);
+        let Some(footer_start) = suffix_len.checked_sub(8) else {
```

This change looks good to me initially, but I'm wondering about the logic behind it.

First of all, it should fail at `file_size < 8`, right? We need to clarify this in our API documentation regarding the validation of `file_size`.

Additionally, it's possible that users provide an incorrect file size. The result of `fetch.fetch(footer_start..file_size).await?;` seems incorrect. We should specify that `fetch` must handle this correctly, returning an error if the range of `footer_start..file_size` fails to read. We can add a check here: if `suffix_len != file_size - footer_start`, we should raise an error like `EOF`.

What do you think?

---

Maybe it is also possible to hit this error if `size_hint` is smaller than 8? A minimum value for `size_hint` could make sense, or returning an error if it is too small.

---

ye I think @jhorstmann was right, the error almost certainly happened because the `prefetch` value was invalid, not because `fetch` returned the wrong size of slice.

I've updated to check `prefetch` size and added a test.